### PR TITLE
CSCFAIRMETA-461: Qvain Light add project

### PR DESCRIPTION
**Pull request by vaahtokarkki**
Add project field to Qvain Light. A dataset can have zero or more projects. A project has details, one or more producer organizations and optional funding agency.

This pr implements also a reusable organization select component, with up to three levels of organizations.
**STATUS:**
- `0...n` projects per dataset, full CRUD
- Project details full CRUD
- Project organizations (`1...n`), full CRUD
- Project funding agencies (`0...n`) full CRUD
- Project funding agencies contributor types (`0...n`) full CRUD

**TODO:**
- Translations

### CODE REVIEW

**Comment by MattiasLevlin** [comment]
Checked things out locally. The organization selection looks great so far, similar to how it's done in actors. The producer project fits well next to the other "advanced" selections (temporal/spatial coverage and related material/history). Nice!

**Comment by vaahtokarkki** [comment]
UI should be tested and reviewed, is there some actions which cause crash or other bugs. It would be good also to double check is the project data saved correctly to Metax and retrieved correctly when editing a dataset containing a project. Finnish translations are not finalized yet.

**Review by ismotee — CHANGES_REQUESTED** [review]
bugs:
- adding project doesn't clear funder type.
- clear button doesn't clear funder type. (Assuming these two to be the same bug)
- adding second project replaces earlier project instead of adding
- Contributor role select field cannot be cancelled

Ui/Ux notes:
- After title and description, when there's no existing projects, there should be text indicating that no projects have been added. Otherwise it's hard to find where freshly created project has gone.
- Clicking on added organization tag to open edit dialog is not obvious (it feels light though). Pen icon is used throughout the application. Now that the icon isn't there, the user tries to find it with no luck. Adding pen icon but keeping the whole tag as an expander of edit fields could work.
- adding and editing organization looks too similar. The user will get confused if, she's adding org and then clicks any of the existing organizations. 
- Clicking on (org or funding agency) tag will possibly erase data that user typed something to add org fields.
- Same goes to project edit button vs. add project. Also existing project remove button clears current add project fields (except funder type)
- because of the nested fields it's quite hard to keep up with the state (add or edit) of the funding agency. For example if the user clicks to edit existing funding agency, then clicks the existing contributor type to edit it and adds something there. Now there's quite a chance that the user clicks edit agency without clicking edit contributor type first. Edited contributor type will be lost. It consistently happens in bigger scale when editing project. One way to solve this is that outer edit button would save all the inner edit states.

add project replacing the existing project blocks testing further.

**Comment by vaahtokarkki** [comment]
Thanks for the feedback! I could reproduce the adding a project replacing the earlier project instead of adding when
- Add project
- Edit project (instead of adding new)
- Clear fields
- Add new --> replaces the previous edited

And that is something what needs to be fixed, good cath!

The issue about funder type not clearing is paused probably by value for react select being undefined instead of null. Undefined value on react select does not clear the select, null will do that. I have fixed that already with the other select fields, so should no take too long to fix.

**Review by ismotee — CHANGES_REQUESTED** [review]
- project without funder type still fails when submitting.This time etsin backend validation seems to reject funder type:
```
WARNING in qvain_light_resources_v2.py:256: Invalid form data: {'projects': {0: {'details': {'funderType': ['Invalid value.']}}}}
```
- project fields are cleared when removing existing project

- Clicking edit button gives controlled to uncontrolled warning from one of the input fields

File: etsin_finder/frontend/locale/finnish.js [hunk 1]
```diff
@@ -620,6 +637,93 @@ const finnish = {
...
+            organization: 'Valitse organisaatio',
+            department: 'Valitse osasto',
+          },
+          addButton: 'Lisää organisaatio',
+          editButton: 'Muokkaa organisaatiota',
+          validation: 'Vähintään yksi organisaatio tarvitaan'
+        },
+        fundingAgency: {
+          contributorType: {
+            title: 'Organisation rooli',
```
Comment (ismotee) [inline etsin_finder/frontend/locale/finnish.js:703]:
```suggestion
            title: 'Organisaation rooli',
```

File: etsin_finder/frontend/locale/finnish.js [hunk 2]
```diff
@@ -620,6 +637,93 @@ const finnish = {
...
+        },
+        titleEn: {
+          placeholder: 'Nimi (Engalnti)',
+        },
+        titleFi: {
+          placeholder: 'Nimi (Suomi)',
+        },
+        identifier: {
+          label: 'Yksilöivä tunniste',
+          description: 'Recommended best practice is to identify the resource by means of a string conforming to a formal identification system. An unambiguous reference to the resource within a given context.',
```
Comment (ismotee) [inline etsin_finder/frontend/locale/finnish.js:678]: "Yksiselitteinen viittaus resurssiin tietyssä kontekstissa. On suositeltavaa käyttää virallisen tunnistamisjärjestelmän mukaisesta tunnistetta."

Näin vapaasti suomennettuna. Placeholderihan pitääkin olla niin kamala että sen haluaa vaihtaa mahdollisimman nopeasti ;)

**Review by ismotee — CHANGES_REQUESTED** [review]
Hyvältä näyttää! Koodi on yleisilmeeltään luettavaa ja siistiä. Muutamat komponentit ovat kohtalaisen isoja ja voisivat olla pilkotumpia. Mikään komponentti ei kuitenkaan ole ylitsepääsemättömän suuri joten en ruvennut niitä erikseen osoittella.

Koodin toisteisuutta oli vähän ja lähinnä jsx:ssä missä se on lähes oletettavaa.

Muutamia ehdotuksia lisäsin koodiin. Merkkasin nit:llä sellaisia huomioita jotak ei tarvi mielestäni juuri nyt korjauksia, mutta jotka herätti tarvetta kommentoida. Osa niistä tuskin tulee aiheuttamaan harmaita hiuksia tulevaisuudessa tai sitten ne vaan ovat toisenlainen tapa tehdä kyseinen ratkaisu. 

Poikkeuksena kuitenkin `project/index.js` kommentti missä huomioin monimutkaisen staten. Se mielestäni pitäisi käsitellä mobx:ssä, mutta staten siirtäminen tässä vaiheessa mobx:ään olisi sen verran iso homma ettei sitä kannata nyt tehdä.

**Comment by vaahtokarkki** [comment]
> Muutamia ehdotuksia lisäsin koodiin. Merkkasin nit:llä sellaisia huomioita jotak ei tarvi mielestäni juuri nyt korjauksia, mutta jotka herätti tarvetta kommentoida. Osa niistä tuskin tulee aiheuttamaan harmaita hiuksia tulevaisuudessa tai sitten ne vaan ovat toisenlainen tapa tehdä kyseinen ratkaisu.
> 
> Poikkeuksena kuitenkin `project/index.js` kommentti missä huomioin monimutkaisen staten. Se mielestäni pitäisi käsitellä mobx:ssä, mutta staten siirtäminen tässä vaiheessa mobx:ään olisi sen verran iso homma ettei sitä kannata nyt tehdä.

Kiitos! Teen vielä noita refaktorointeja/muutoksia mitä ehdotit. Mobx luokan tekeminen formin tilan hallintaan voisi olla yksi vaihtoehto, etenkin kun lomake paisui oletettua monimutkaisemmaksi. Tämä on myös ehkä arkitehtuurikysymys, itse en lähtökohtaisesti näe tarpeellisena säilyttää yksittäisen (ali)lomakkeen tilaa globaalissa tilassa, jos sille ei mitään erityistä syytä ole (tässä monimutkainen state voisi ehkä olla sellainen).

File: etsin_finder/frontend/js/components/qvain/general/organizationSelect.jsx [hunk 3]
```diff
@@ -0,0 +1,498 @@
...
+      </SelectContainer>
+    )
+  }
+}
+
+/**
+ * Internally used select component with form for adding organization manually.
+ * This is a stateless component.
+ */
+const CreatableSelectComponent = ({ Stores, onChange, onBlur, value, options, placeholder, name, inputId, creatable, allowReset }) => {
```
[the remark this answers is not in the document]
Comment (vaahtokarkki) [reply etsin_finder/frontend/js/components/qvain/general/organizationSelect.jsx:249]: This component (as well some other similar cases in this pr) is a private component of the exported "main" component. That is why it is not in own file, it should not be exported to make sure that no other components uses it. 
Downside of this is that the private component bloats the file of parent component.

File: etsin_finder/frontend/js/components/qvain/project/index.jsx [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+      organizations: { ...INITIAL_STATE.organizations, addedOrganizations: [...organizations] },
+      fundingAgencies: { ...INITIAL_STATE.fundingAgencies, addedFundingAgencies: [...fundingAgencies] },
+      projectInEdit: true,
+    })
+  }
+
+  removeProject = (id, event) => {
+    if (this.props.Stores.Qvain.readonly) return
+    if (event) event.preventDefault()
+    this.resetForm()
```
[the remark this answers is not in the document]
Comment (vaahtokarkki) [reply etsin_finder/frontend/js/components/qvain/project/index.jsx:183]: Fixed

**Review by ismotee — CHANGES_REQUESTED** [review]
Found few new bugs that I missed last round:
- edit button doesn't clear existing values on fields that are empty in project that is to be edited
   - this bug is related to controlled to uncontrolled warning that appear when clicking two different edit button
- application crashes when funder type field is selected and trying to type something that doesn't match any option
warning:
```
Warning: Cannot update during an existing state transition (such as within `render`). Render methods should be a pure function of props and state.
```
error:
```
Uncaught TypeError: noOptionsMessage is not a function
```

**Pull request merged**